src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java

```diff
@@ -420,79 +481,165 @@ public static long getEffectivePermission(Channel channel, Role role)
-                }
-                else
-                {
-                    allow = po.getAllowedRaw() | allow;     //Give all the stuff allowed by this Role's allow
-
-                    deny = (po.getDeniedRaw() | deny) & (~allow);  //Deny everything that this role denies.
-                    // This also rewrites the previous role's denies if this role allowed those permissions.
-                }
+                allow = override.getAllowedRaw();
+                deny = override.getDeniedRaw();
+            }
+            else
+            {
+                allow |= override.getAllowedRaw();
+                deny = (override.getDeniedRaw() | deny) & (~allow);
             }
         }
-        if (allow != -1 && deny != -1)  //If we found at least 1 role with overrides.
-        {
+
+        if (allow >= 0 && deny >= 0)
             permission = apply(permission, allow, deny);
-        }
 
-        //handle member-specific overrides
-        PermissionOverride memberOverride = memberOverrides.get(member);
-        if (memberOverride != null)
-        {
-            permission = apply(permission, memberOverride.getAllowedRaw(), memberOverride.getDeniedRaw());
-        }
+        override = channel.getPermissionOverride(member);
+        if (override != null)
+            permission = apply(permission, override.getAllowedRaw(), override.getDeniedRaw());
+
         return permission;
     }
 
+    /**
+     * Retrieves the implicit permissions of the specified {@link net.dv8tion.jda.core.entities.Role Role}
+     * in its hosting {@link net.dv8tion.jda.core.entities.Guild Guild} and specific {@link net.dv8tion.jda.core.entities.Channel Channel}.
+     * <br><b>Allowed permissions override denied permissions of {@link net.dv8tion.jda.core.entities.PermissionOverride PermissionOverrides}!</b>
+     *
+     * <p>All permissions returned are explicitly granted to this Role.
+     * <br>Permissions like {@link net.dv8tion.jda.core.Permission#ADMINISTRATOR Permission.ADMINISTRATOR} do not
+     * grant other permissions in this value.
+     * <p>This factor in existing {@link net.dv8tion.jda.core.entities.PermissionOverride PermissionOverrides} if possible.
+     *
+     * @param  channel
+     *         The target channel of which to check {@link net.dv8tion.jda.core.entities.PermissionOverride PermissionOverrides}
+     * @param  role
+     *         The non-null {@link net.dv8tion.jda.core.entities.Role Role} for which to get implicit permissions
+     *
+     * @throws IllegalArgumentException
+     *         If any of the arguments is {@code null}
+     *         or the specified entities are not from the same {@link net.dv8tion.jda.core.entities.Guild Guild}
+     *
+     * @return Primitive (unsigned) long value with the implicit permissions of the specified role in the specified channel
+     *
+     * @since  3.1
+     */
+    public static long getImplicitPermission(Channel channel, Role role)
```

Comment:
That is only dealing with the override information. I don't see you ever taking into consideration the `role.getPermissionRaw`.